x/programs/rust/wasmlanche_sdk/src/simulator.rs

```diff
@@ -0,0 +1,363 @@
+pub enum Endpoint {
+    /// Perform an operation against the key api.
+    Key,
+    /// Make a read-only call to a program function and return the result.
+    View,
+    /// Create a transaction from a possible state changing program function
+    /// call. A program's function can internally optionally call other
+    /// functions including program to program.
+    Execute,
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+pub struct Step<'a> {
+    /// A description of the step.
+    description: &'a str,
+    /// The API endpoint to call.
+    endpoint: Endpoint,
+    /// The method to call on the endpoint.
+    method: &'a str,
+    /// The parameters to pass to the method.
+    params: Vec<Param<'a>>,
+    #[serde(default)]
+    require: Option<Require>,
+}
+
+#[derive(Debug, PartialEq)]
+pub enum ParamType {
+    U64,
+    String,
+    Key(Key),
+    ID,
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+#[serde(rename_all = "lowercase")]
+pub enum Key {
+    Ed25519,
+    Secp256r1,
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+pub struct Param<'a> {
+    /// The optional name of the parameter. This is used for readability.
+    name: &'a str,
+    #[serde(rename = "type")]
+    /// The type of the parameter.
+    param_type: ParamType,
+    /// The value of the parameter.
+    value: &'a str,
+}
+
+impl Serialize for ParamType {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        match self {
+            ParamType::U64 => serializer.serialize_str("u64"),
+            ParamType::String => serializer.serialize_str("string"),
+            ParamType::Key(Key::Ed25519) => serializer.serialize_str("ed25519"),
+            ParamType::Key(Key::Secp256r1) => serializer.serialize_str("secp256r1"),
+            ParamType::ID => serializer.serialize_str("id"),
+        }
+    }
+}
+
+impl<'de> Deserialize<'de> for ParamType {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        use serde::de::Error;
+        let s = String::deserialize(deserializer)?;
+        match s.as_str() {
+            "u64" => Ok(ParamType::U64),
+            "string" => Ok(ParamType::String),
+            "ed25519" => Ok(ParamType::Key(Key::Ed25519)),
+            "secp256r1" => Ok(ParamType::Key(Key::Secp256r1)),
+            "id" => Ok(ParamType::ID),
+            _ => Err(D::Error::custom(format!("unknown param type: {s}"))),
+        }
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+pub struct Require {
+    result: ResultCondition,
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+pub enum Operator {
+    #[serde(rename = "==")]
+    NumericEq,
+    #[serde(rename = "!=")]
+    NumericNe,
+    #[serde(rename = ">")]
+    NumericGt,
+    #[serde(rename = "<")]
+    NumericLt,
+    #[serde(rename = ">=")]
+    NumericGe,
+    #[serde(rename = "<=")]
+    NumericLe,
+}
```

Comment:
Going to expand this in the future?
If not
```suggestion
#[derive(Debug, Serialize, Deserialize, PartialEq)]
pub enum NumericOperator {
    #[serde(rename = "==")]
    Eq,
    #[serde(rename = "!=")]
    Ne,
    #[serde(rename = ">")]
    Gt,
    #[serde(rename = "<")]
    Lt,
    #[serde(rename = ">=")]
    Ge,
    #[serde(rename = "<=")]
    Le,
}
```

Reply:
Yeah I was THINKING of strings in the future but I'm going to use this

Reply:
going to leave it for now actually.